### DISCUSSION TITLE
Fetching slug from MongoDB collection returning undefined - App router

### DISCUSSION THREAD

**Discussion by evangordo** [vercel/next.js#65214] [Help]
### Summary

**Update: Edited with added error handling**

I'm trying to get the individual slug for each job that I have posted from a MongoDB collection. I  have set up the following page: `src/app/Job/[slug]/page.js`

Here's the code:

```
export async function getPost(slug) {
  const url = `http://localhost:3000/api/jobs/${slug}`;
  try {
    const res = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!res.ok) {
      throw new Error(`HTTP error! Status: ${res.status} for URL: ${url}`);
    }
    return await res.json();
  } catch (error) {
    console.error("failed to fetch post:", error);
    throw error;
  }
}

export default async function Post({ params }) {
  const post = await getPost(params);

  return (
    <>
      <h1>{post.pay}</h1>
      <h1>{post.aboutCompany}</h1>
      <h1>{post.experience}</h1>
    </>
  );
}

```

I have several jobs posted, but when I attempt to click on one using this `<Link href={/Job/${slug}}>View job</Link>`, 

i get **Error: HTTP error! Status: 404 for URL: http://localhost:3000/api/jobs/undefined**

Any help be much appreciated!

**Comment by icyJoseph** [comment]
There's a couple of things here...

- Lack of error handling
- Reaching for API routes, when you are already within a server component, [more about that here](https://nextjs-faq.com/fetch-api-in-rsc)
- And the main error, passing `params` which is an object, but expecting it to be a string

```ts
export default async function Post({ params }) {
  const post = await getPost(params); // params here is something like { slug: foo }
  /* rest of implementation */
```

So that means, you have to change either to `getPost(params.slug)` or:

```ts
async function getPost({ slug }) {
/* rest of implementation */
}

```

You can find all of these on the docs, https://nextjs.org/docs/app/api-reference/file-conventions/page

**Reply by evangordo** [reply]
Thank you for the response! I've implemented your feedback, and it has been helpful. I now realise that my **slug URL is undefined**. Additionally, thanks for the article. I'm just testing route in my component to try get my slug working for now. I will refactor it once I figure out this bug.

```
export async function getPost(slug) {
  const url = `http://localhost:3000/api/jobs/${slug}`;
  console.log("Requesting URL:", url);
  try {
    const res = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!res.ok) {
      throw new Error(`HTTP error! Status: ${res.status} for URL: ${url}`);
    }
    return await res.json();
```

**Reply by icyJoseph** [reply]
So slug is still undefined? How are you calling this function? Are you passing params.slug then?

**Reply by evangordo** [reply]
yes

```
export default async function Page({ params }) {
  const post = await getPost(params.slug);

  return (
    <div>
      <h1>{post.aboutCompany}</h1>
      <p>{post.pay}</p>
    </div>
  );
}
```
